arch/risc-v/src/common/riscv_misaligned.c

```diff
@@ -0,0 +1,448 @@
+              insn.lwsp.imm76 << 6;
+        ctx->dest = (uint8_t *)&regs[insn.lwsp.rd];
+        ctx->src  = (uint8_t *)regs[REG_SP] + imm;
+        ctx->len  = 4;
+        break;
+
+      case INSN_C_LD:
+        imm = insn.ld.imm53 << 3 | insn.ld.imm76 << 6;
+        ctx->dest = (uint8_t *)&regs[REG_X8 + insn.ld.rd];
+        ctx->src  = (uint8_t *)regs[REG_X8 + insn.ld.rs1] + imm;
+        ctx->len  = 8;
+        break;
+
+      case INSN_C_LDSP:
+        imm = insn.ldsp.imm43 << 3 | insn.ldsp.imm5 << 5 |
+              insn.ldsp.imm86 << 6;
+        ctx->dest = (uint8_t *)&regs[insn.ld.rd];
+        ctx->src  = (uint8_t *)regs[REG_SP] + imm;
+        ctx->len  = 8;
+        break;
+
+      case INSN_C_SW:
+        imm = insn.sw.imm2 << 2 | insn.sw.imm53 << 3 | insn.sw.imm6 << 6;
+        ctx->dest = (uint8_t *)regs[REG_X8 + insn.sw.rs1] + imm;
+        ctx->src  = (uint8_t *)&regs[REG_X8 + insn.sd.rs2];
+        ctx->len  = 4;
+        break;
+
+      case INSN_C_SWSP:
+        imm = insn.swsp.imm52 << 2 | insn.swsp.imm76 << 6;
+        ctx->dest = (uint8_t *)regs[REG_SP] + imm;
+        ctx->src  = (uint8_t *)&regs[insn.swsp.rs2];
+        ctx->len  = 4;
+        break;
+
+      case INSN_C_SD:
+        imm = insn.sd.imm53 << 3 | insn.sd.imm76 << 6;
+        ctx->dest = (uint8_t *)regs[REG_X8 + insn.sd.rs1] + imm;
+        ctx->src  = (uint8_t *)&regs[REG_X8 + insn.sd.rs2];
+        ctx->len  = 8;
+        break;
+
+      case INSN_C_SDSP:
+        imm = insn.sdsp.imm53 << 3 | insn.sdsp.imm86 << 6;
+        ctx->dest = (uint8_t *)regs[REG_SP] + imm;
+        ctx->src  = (uint8_t *)&regs[insn.sdsp.rs2];
+        ctx->len  = 8;
+        break;
+
+      default:
+        _alert("Compressed: %x\n", insn.insn);
+        return false;
+    }
+
+  regs[REG_EPC] += 2;
+
+  return true;
+#else
+  return false;
+#endif
+}
+
+/****************************************************************************
+ * Name: decode_insn_ctx
+ *
+ * Description:
+ *   Try to decode a normal encoding
+ *
+ ****************************************************************************/
+
+static bool decode_insn_ctx(uintptr_t *regs, rv_insn_ctx_t *ctx)
+{
+  uint32_t in;
+  rv_insn_t insn;
+  int32_t imm;
+
+  insn.insn = *(uint32_t *)regs[REG_EPC];
+
+  /* Get load/store instruction encoding */
+
+  in = insn.insn & INSN_MASK;
+
+  switch (in)
+    {
+      case INSN_LH:
+      case INSN_LHU:
+      case INSN_LW:
+      case INSN_LWU:
+      case INSN_LD:
+
+        /* Load a value from memory to register */
+
+        ctx->dest = (uint8_t *)&regs[insn.l.rd];
+        ctx->src = (uint8_t *)regs[insn.l.rs1] + insn.l.imm;
+
+        /* Zero the target register */
+
+        *(uintptr_t *)ctx->dest = 0;
+
+        /* Get data wide bit */
+
+        in &= (~0x4000);
+        in >>= 12;
+        ctx->len = 2 << (in - 1);
+
+        break;
+
+      case INSN_SH:
+      case INSN_SW:
+      case INSN_SD:
+
+        /* Fetch sign extended offset */
```

Comment:
where is the sign extension

Reply:
Emm, it just means the immediate value is a signed value